### ISSUE TITLE
Add Support for Azure VirtualWAN

### ISSUE THREAD

**Issue by Theragus** [infracost/infracost#910]
Terraform Resources: 

`azurerm_virtual_wan`
`azurerm_virtual_hub`

Pricing Information: 

https://azure.microsoft.com/en-us/pricing/details/virtual-wan/

## azurerm_virtual_hub

Terraform resource: https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/virtual_hub

Cost components:

* Deployment - unit `hours`
* Data processing - unit `GB`

Notes:

* If `type` is `Basic` then we should treat it as a free resource

Usage file:

```yml
  azurerm_virtual_hub.my_virtual_hub:
    monthly_data_processed_gb # Monthly data processed by the Virtual WAN Hub in GB
```

## azurerm_vpn_gateway

Cost components:

* `S2S scale units` - unit `scale units`. Quantity comes from `scale_units` on Terraform resource.

## azurerm_point_to_site_vpn_gateway

Cost components:

* `P2S scale units` - unit `scale units`. Quantity comes from `scale_units` on Terraform resource.
* `P2S connections` - unit `connections`. Hourly quantity is set to 0. Usage parameter `monthly_p2s_connections_hrs` overrides this.

## azurerm_express_route_gateway

Cost components:

* `ER scale units` - unit `scale units`. Quantity comes from `scale_units` on Terraform resource.

 azurerm_point_to_site_vpn_gateway, azurerm_express_route_gateway
 
Note:
* This resource has the prices from https://azure.microsoft.com/en-us/pricing/details/virtual-wan/, not https://azure.microsoft.com/en-gb/pricing/details/expressroute/. It seems the skus for the ExpressRoute Gateway resource there are handled by this resource https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/virtual_network_gateway

## azurerm_vpn_gateway_connection

Cost components:

* `S2S connections` - unit `hours`. Hourly quantity is set to 1.

## azurerm_express_route_connection

Cost components:

* `ER connections` - unit `hours`. Hourly quantity is set to 1.

Usage file:

* No usage file parameters needed

## Free resources

* `azure_virtual_wan` - since all the costs are covered in the `azure_virtual_hub` resource.

Unsupported:

* All remaining `azurerm_express_route_*` - This has [other prices that apply as well](https://azure.microsoft.com/en-gb/pricing/details/expressroute/). When we add this resource we can also add the "Connection" and "Scale unit" costs.
* NVA Infrastructure Unit and Routing Infrastructure Unit - I can't find any Terraform resources that map to these.

**Comment by aliscott** [comment]
I see as well as deployment and data processing prices, there's also prices for connections and scale units. I'm not familiar with these Azure resources, so here's a suggestion for how we could break this down. @Theragus please add any suggestions you have for how to match these to the Terraform resources.

## azurerm_virtual_hub

Terraform resource: https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/virtual_hub

Cost components:

* Deployment - unit `hours`
* Data processing - unit `GB`

Notes:

* If `type` is `Basic` then we should treat it as a free resource

Usage file:

```yml
  azurerm_virtual_hub.my_virtual_hub:
    monthly_data_processed_gb # Monthly data processed by the Virtual WAN Hub in GB
```

## azurerm_vpn_gateway, azurerm_point_to_site_vpn_gateway

Cost components:

* Scale units - unit `scale units`

Notes:

* There's also [`azurerm_virtual_network_gateway`](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/data-sources/virtual_network_gateway) but confusingly this seems to be for [these prices](https://azure.microsoft.com/en-gb/pricing/details/vpn-gateway/)

Usage file:

* No usage file parameters needed

## azurerm_vpn_gateway_connection

Cost components:

* Connection (point-to-site/site-to-site) - unit `hours`

Notes:

* Cost depends on if the `vpn_gateway_id` relates to a `azurerm_vpn_gateway` (site-to-site) or a `azurerm_point_to_site_vpn_gateway` (point-to-site)

Usage file:

* No usage file parameters needed

## Free resources

* `azure_virtual_wan` - since all the costs are covered in the `azure_virtual_hub` resource.

Unsupported:

* `azurerm_express_route_*` - This has [other prices that apply as well](https://azure.microsoft.com/en-gb/pricing/details/expressroute/). When we add this resource we can also add the "Connection" and "Scale unit" costs.
* NVA Infrastructure Unit and Routing Infrastructure Unit - I can't find any Terraform resources that map to these.

**Comment by Theragus** [comment]
Hi @aliscott, sorry for the late reply, I already read your comment a couple of weeks ago, but somehow it went completely over my head that feedback was requested here :/

I now managed to test this extensively and look into a deployment that I already set up months ago.

# azurerm_virtual_hub

looks perfectly fine to me, I had to do a little bit of digging inside the MS docs to confirm that basic SKU incurs no charges since it's not mentioned on the pricing page or in the pricing calculator. I found the only information here https://docs.microsoft.com/en-us/azure/virtual-wan/pricing-concepts#fee

# azurerm_vpn_gateway, azurerm_point_to_site_vpn_gateway, azurerm_express_route_connection

This is now where stuff starts to get REALLY tricky. Basically, Virtual WAN has 3 distinct connection types.

- Site-to-Site VPN connections (`azurerm_vpn_gateway`, `azurerm_vpn_gateway_connection`, `azurerm_vpn_site`)

- Point-to-Site VPN connections (`azurerm_vpn_server_configuration`, `azurerm_point_to_site_vpn_gateway`)

- ExpressRoute connections (`azurerm_express_route_connection`, `azurerm_express_route_gateway`, `azurerm_express_route_port`)

I wrote the corresponding terraform resources to every connection type. 

Notice that `azurerm_vpn_server_configuration` has nothing to do with Site-to-Site VPN, as well as `azurerm_express_route_*` having nothing to do with `azurerm_express_route_circuit_*`, same goes for `azurerm_virtual_network_gateway` having nothing to do with `azurerm_vpn_gateway`

MS is incurring charges for all 3 of these connection types. But, luckily for us, neither is the same as the other!

They all incur charges after connection units (number of S2S, P2S, and ER connections) as well as so-called scaling units, which determine the amount of bandwidth available. P2S and S2S scaling units scale in increment of 500 Mbps, while ER scales in increment of 2 Gbps.

I now tried to tie the specific resources to the incurred charges below, so I hope this helps in understanding where what is billed. (https://azure.microsoft.com/en-us/pricing/details/virtual-wan/)

`azurerm_vpn_gateway` = S2S scaling unit
`azurerm_vpn_gateway_connection` = S2S connection unit
`azurerm_point_to_site_vpn_gateway` = P2S scaling unit
`azurerm_express_route_gateway` = ER scaling unit
`azurerm_express_route_connection` = ER connection unit


Note: P2S has no fixed connection units since it varies based on how many users are connecting to the VPN Gateway.

There's also the Secured Virtual WAN with Firewalls pricing, but since the azurerm provider doesn't support it anyway (for now currently) it makes no sense thinking about that. 
Also, this was already long enough to write so I think there's enough to grasp here...

**Comment by hugorut** [comment]
Update for visibility - I made a start on this last weekend and planned to have it in PR around now, but work has been hectic. I'm hopeful to get some time to work on it tomorrow eve.

**Comment by alikhajeh1** [comment]
@hugorut heads-up that @vdmgolub is working on https://github.com/infracost/infracost/issues/909 and I think he's putting in the scaffolding for Azure resources to use the new method of adding resources as documented in the contributing guide. You might want to wait for him to send his PR to avoid conflicts 🤷

**Comment by hugorut** [comment]
lads, @aliscott/ @alikhajeh1 I've failed miserably at making a dent in this issue. I just haven't found the time after work to give it bandwidth, so I only ever reached the scoping stages. I'm off on holiday tomorrow and won't be back in action until the 22nd. So I'm going to unassign myself and put it back on the stack in case anyone else wants to start it whilst I'm away. I can pick it up again on the 22nd If no-one else has taken a punt on it.

**Comment by aliscott** [comment]
🙌  thanks Hugo, have a good holiday!

**Comment by aliscott** [comment]
FYI I've updated the description of this issue based on the feedback from @Theragus.

**Comment by hugorut** [comment]
@alikhajeh1 alrighty - back from hols and now have some bandwidth. All good if I pick this up again?

**Comment by alikhajeh1** [comment]
Sure - thanks! Let us know if you need test Azure creds to run TF plan etc.

**Comment by alikhajeh1** [comment]
@Theragus just FYI that this will be released in Infracost CLI v0.9.15 by next week.

**Comment by aliscott** [comment]
@Theragus this has now been released in v0.9.5!
